src/data/setoid/partition.lean

```diff
@@ -201,3 +212,123 @@ _ (subtype (@is_partition α)) _ (partial_order.to_preorder _) $ partition.rel_i
 end partition
 
 end setoid
+
+/-- A partition of a type `α` indexed by another type `ι`. -/
+@[nolint has_inhabited_instance]
+structure indexed_partition {ι α : Type*} (s : ι → set α) :=
+(eq_of_mem : ∀ {x i j}, x ∈ s i → x ∈ s j → i = j)
+(some : ι → α)
+(some_mem : ∀ i, some i ∈ s i)
+(index : α → ι)
+(mem_index : ∀ x, x ∈ s (index x))
+
+/-- The non-constructive constructor for `indexed_partition`. -/
+noncomputable
+def indexed_partition.mk' {ι α : Type*} (s : ι → set α) (dis : ∀ i j, i ≠ j → disjoint (s i) (s j))
+(empty : ∀ i, (s i).nonempty) (ex : ∀ x, ∃ i, x ∈ s i) : indexed_partition s :=
+{ eq_of_mem := begin
+    classical,
+    intros x i j hxi hxj,
+    by_contra h,
+    exact dis _ _ h ⟨hxi, hxj⟩
+  end,
+  some := λ i, (empty i).some,
+  some_mem := λ i, (empty i).some_spec,
+  index := λ x, (ex x).some,
+  mem_index := λ x, (ex x).some_spec }
+
+namespace indexed_partition
+
+open set
+
+variables {ι α : Type*} {s : ι → set α} (hs : indexed_partition s)
+include hs
+
+lemma exists_mem (x : α) : ∃ i, x ∈ s i := ⟨hs.index x, hs.mem_index x⟩
+
+lemma Union : (⋃ i, s i) = univ :=
+by { ext x, simp [hs.exists_mem x] }
+
+lemma disjoint : ∀ {i j}, i ≠ j → disjoint (s i) (s j) :=
+λ i j h x ⟨hxi, hxj⟩, h (hs.eq_of_mem hxi hxj)
+
+lemma mem_iff_index_eq {x i} : x ∈ s i ↔ hs.index x = i :=
+⟨λ hxi, (hs.eq_of_mem hxi (hs.mem_index x)).symm, λ h, h ▸ hs.mem_index _⟩
+
+lemma eq (i) : s i = {x | hs.index x = i} :=
+set.ext $ λ _, hs.mem_iff_index_eq
+
+/-- The equivalence relation associated to an indexed partition. Two
+elements are equivalent if they belong to the same set of the partition. -/
+protected abbreviation setoid (hs : indexed_partition s) : setoid α :=
+setoid.ker hs.index
+
+@[simp] lemma index_some (i : ι) : hs.index (hs.some i) = i :=
+(mem_iff_index_eq _).1 $ hs.some_mem i
+
+lemma some_index (x : α) : hs.setoid.rel (hs.some (hs.index x)) x :=
+hs.index_some (hs.index x)
+
+/-- The quotient associated to an indexed partition. -/
+@[nolint has_inhabited_instance]
```

Comment:
I think it's pretty easy to make an inhabited instance here by just applying `proj` to an arbitrary element of `α`

Reply:
What `s` would you choose for the instance?

Reply:
Sure - I don't see why not!

Reply:
Done